front/lib/api/assistant/email_trigger.ts

```diff
@@ -0,0 +1,486 @@
+
+function renderUserType(user: User): UserType {
+  return {
+    sId: user.sId,
+    id: user.id,
+    createdAt: user.createdAt.getTime(),
+    provider: user.provider,
+    username: user.username,
+    email: user.email,
+    firstName: user.firstName,
+    lastName: user.lastName,
+    fullName: user.firstName + (user.lastName ? ` ${user.lastName}` : ""),
+    image: user.imageUrl,
+  };
+}
+
+export const ASSISTANT_EMAIL_SUBDOMAIN = isDevelopment()
+  ? "run.dust.help"
+  : "run.dust.help";
+
+export type InboundEmail = {
+  subject: string;
+  text: string;
+  auth: { SPF: string; dkim: string };
+  envelope: {
+    to: string[];
+    cc: string[];
+    bcc: string[];
+    from: string;
+    full: string;
+  };
+};
+
+export type EmailTriggerError = {
+  type:
+    | "unexpected_error"
+    | "unauthenticated_error"
+    | "user_not_found"
+    | "workspace_not_found"
+    | "invalid_email_error"
+    | "assistant_not_found"
+    | "message_creation_error";
+  message: string;
+};
+
+export function getTargetEmailsForWorkspace({
+  allTargetEmails,
+  workspace,
+  isDefault,
+}: {
+  allTargetEmails: string[];
+  workspace: LightWorkspaceType;
+  isDefault: boolean;
+}): string[] {
+  return allTargetEmails.filter(
+    (email) =>
+      email.split("@")[0].endsWith(`[${workspace.sId}]`) ||
+      // calls with no brackets go to default workspace
+      (!email.split("@")[0].endsWith("]") && isDefault)
+  );
+}
+
+export async function userAndWorkspacesFromEmail({
+  email,
+}: {
+  email: string;
+}): Promise<
+  Result<
+    {
+      workspaces: LightWorkspaceType[];
+      user: UserType;
+      defaultWorkspace: LightWorkspaceType;
+    },
+    EmailTriggerError
+  >
+> {
+  const user = await User.findOne({
+    where: { email },
+  });
+
+  if (!user) {
+    return new Err({
+      type: "user_not_found",
+      message:
+        `Failed to match a valid Dust user for email: ${email}. ` +
+        `Please sign up for Dust at https://dust.tt to interact with assitsants over email.`,
+    });
+  }
+  const workspaces = await Workspace.findAll({
+    include: [
+      {
+        model: MembershipModel,
+        where: {
+          userId: user.id,
+          endAt: {
+            [Op.or]: [{ [Op.is]: null }, { [Op.gte]: new Date() }],
+          },
+        },
+      },
+    ],
+  });
+
+  if (!workspaces) {
+    return new Err({
+      type: "workspace_not_found",
+      message:
+        `Failed to match a valid Dust workspace associated with email: ${email}. ` +
+        `Please sign up for Dust at https://dust.tt to interact with assistants over email.`,
+    });
+  }
+
+  // get latest conversation participation from user
+  const latestParticipation = await ConversationParticipant.findOne({
+    where: {
+      userId: user.id,
+    },
+    include: [
+      {
+        model: Conversation,
+      },
+    ],
+    order: [["createdAt", "DESC"]],
+  });
+
+  // TODO: implement good default logic
```

Comment:
Can we pin it to our workspace for now?

Reply:
That way we can remove the call to ConversationParticipants and make it super simple?

Reply:
totes :+1: